File: architecture-skills/design-patterns/examples/singleton/singleton.go
// Singleton pattern in Go.
//
// sync.Once is the standard form: the accessor lazily initializes the
// package-level instance exactly once, safely across goroutines. Prefer
// constructing the config in main and passing it down where feasible;
// package-level singletons make dependencies invisible and tests coupled.

package main

import (
	"fmt"
	"sync"
)

type AppConfig struct {
	settings map[string]string
}

func (c *AppConfig) Get(key string) string {
	return c.settings[key]
}

var (
	instance *AppConfig
	once     sync.Once
)

func GetAppConfig() *AppConfig {
	once.Do(func() {
		// Stands in for reading a config file once.
		fmt.Println("Loading configuration (expensive, happens once)")
		instance = &AppConfig{
			settings: map[string]string{
				"app_name":  "InventoryService",
				"db_url":    "postgres://localhost/inventory",
				"log_level": "INFO",
			},
		}
	})
	return instance
}

func main() {
	a := GetAppConfig()
	b := GetAppConfig()

	fmt.Println("same instance:", a == b)
	fmt.Println("app_name:", a.Get("app_name"))
	fmt.Println("db_url:", b.Get("db_url"))
}

File: architecture-skills/design-patterns/examples/singleton/singleton.rs
// Singleton pattern in Rust.
//
// std::sync::OnceLock initializes a static exactly once, thread-safely
// (LazyLock and the older lazy_static/once_cell crates are equivalent).
// Rust makes the costs explicit: a mutable singleton would need a Mutex.
// Prefer passing &AppConfig down from main where feasible; globals hide
// dependencies from function signatures.

use std::collections::HashMap;
use std::sync::OnceLock;

#[derive(Debug)]
struct AppConfig {
    settings: HashMap<&'static str, &'static str>,
}

impl AppConfig {
    fn load() -> Self {
        // Stands in for reading a config file once.
        println!("Loading configuration (expensive, happens once)");
        Self {
            settings: HashMap::from([
                ("app_name", "InventoryService"),
                ("db_url", "postgres://localhost/inventory"),
                ("log_level", "INFO"),
            ]),
        }
    }

    fn get(&self, key: &str) -> &str {
        self.settings[key]
    }
}

fn app_config() -> &'static AppConfig {
    static INSTANCE: OnceLock<AppConfig> = OnceLock::new();
    INSTANCE.get_or_init(AppConfig::load)
}

fn main() {
    let a = app_config();
    let b = app_config();

    println!("same instance: {}", std::ptr::eq(a, b));
    println!("app_name: {}", a.get("app_name"));
    println!("db_url: {}", b.get("db_url"));
}
